src/app-admin.js

```diff
@@ -0,0 +1,491 @@
+import express from 'express';
+import * as uuid from 'uuid';
+import { errorMiddleware } from './util/middlewares.js';
+import validateUser from './util/validate-user.js';
+import getAccountDb, { isAdmin } from './account-db.js';
+import config from './load-config.js';
+import bodyParser from 'body-parser';
+
+let app = express();
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(errorMiddleware);
+
+export { app as handlers };
+
+export const TOKEN_EXPIRATION_NEVER = -1;
+
+const sendErrorResponse = (res, status, reason, details) => {
+  res.status(status).send({
+    status: 'error',
+    reason,
+    details,
+  });
+};
+
+const getUserByUsername = (userName) => {
+  return (
+    getAccountDb().first('SELECT id FROM users WHERE user_name = ?', [
+      userName,
+    ]) || {}
+  );
+};
+
+const getFileById = (fileId) => {
+  return (
+    getAccountDb().first('SELECT id FROM files WHERE files.id = ?', [fileId]) ||
+    {}
+  );
+};
+
+export const getAdminSessionFromRequest = async (req, res) => {
+  let session = validateUser(req, res);
+  if (!session) {
+    return null;
+  }
+
+  if (!(await isAdmin(session.user_id))) {
+    sendErrorResponse(res, 401, 'unauthorized', 'permission-not-found');
+    return null;
+  }
+
+  return session;
+};
+
+const validateUserInput = (res, user) => {
+  if (!user.userName) {
+    sendErrorResponse(
+      res,
+      400,
+      'user-cant-be-empty',
+      'Username cannot be empty',
+    );
+    return false;
+  }
+
+  if (!user.role) {
+    sendErrorResponse(res, 400, 'role-cant-be-empty', 'Role cannot be empty');
+    return false;
+  }
+
+  const { id: roleId } =
+    getAccountDb().first('SELECT id FROM roles WHERE roles.id = ?', [
+      user.role,
+    ]) || {};
+
+  if (!roleId) {
+    sendErrorResponse(
+      res,
+      400,
+      'role-does-not-exists',
+      'Selected role does not exists',
+    );
+    return false;
+  }
+
+  return true;
+};
+
+app.get('/masterCreated/', (req, res) => {
+  const { cnt } =
+    getAccountDb().first(
+      `SELECT count(*) as cnt
+     FROM users
+     WHERE users.user_name <> '' and users.master = 1`,
+    ) || {};
+
+  res.json(cnt > 0);
+});
+
+app.get('/users/', (req, res) => {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏should non-admins be able to see the full user list?

Reply:
if they have a budget, they need to know the user list to give access. since the route needs a session, it's not totally public, only users from that instance can get user list